lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -55,142 +53,139 @@ def tags(repository, params = {})
 
       # Replaces the user-repo mapping for a single user
-      def update_user_repositories(username, repositories)
-        user = User.where(name: username).first
-        user.remove_all_repositories
-        repositories.each do |repo_name|
-          found_repo = Repository.find(name: repo_name)
-          if found_repo.nil?
-            logger.warn("#{repo_name} does not exist in this smart proxy's environments")
-          elsif user.repositories_dataset.where(name: repo_name).first.nil? && found_repo.auth_required
-            user.add_repository(found_repo)
-          end
+      def update_user_repositories(database, username, repositories)
+        user = database.connection[:users][{ name: username }]
+
+        user_repositories = database.connection[:repositories_users]
+        database.connection.transaction do
+          user_repositories.where(user_id: user[:id]).delete
+
+          user_repositories.import(
+            %i[repository_id user_id],
+            database.connection[:repositories].where(name: repositories, auth_required: true).select(:id).map do |repo|
+              [repo[:id], user[:id]]
+            end
+          )
         end
       end
 
-      def authorized_for_repo?(repo_name, user_token_is_valid, username = nil)
-        repository = Repository.where(name: repo_name).first
+      def authorized_for_repo?(database, repo_name, user_token_is_valid, username = nil)
+        repository = database.connection[:repositories][{ name: repo_name }]
 
         # Repository doesn't exist
         return false if repository.nil?
 
         # Repository doesn't require auth
-        return true unless repository.auth_required
+        return true unless repository[:auth_required]
 
-        if username && user_token_is_valid && repository.auth_required
+        if username && user_token_is_valid
           # User is logged in and has access to the repository
-          user = User.find(name: username)
-          return !user.repositories_dataset.where(name: repo_name).first.nil?
+          return !database.connection[:repositories_users].where(
+            repository_id: repository[:id], user_id: database.connection[:users].first(name: username)[:id]
+          ).empty?
         end
 
         false
       end
 
-      def token_user(token)
-        User[AuthenticationToken.find(token_checksum: Digest::SHA256.hexdigest(token)).user_id]
+      def token_user(database, token)
+        database.connection[:users][{
+          id: database.connection[:authentication_tokens].where(token_checksum: checksum(token)).select(:user_id)
+        }]
       end
 
-      def valid_token?(token)
-        AuthenticationToken.where(token_checksum: Digest::SHA256.hexdigest(token)).where do
+      def valid_token?(database, token)
+        !database.connection[:authentication_tokens].where(token_checksum: checksum(token)).where do
           expire_at > Sequel::CURRENT_TIMESTAMP
-        end.count.positive?
+        end.empty?
```

Comment:
By doing the `EXISTS` check server side, you don't need to transfer all the results over the wire. It may not matter for network traffic since it's all a local file now, but for remote databases it's a common trick. Here I can also imagine it doesn't need to serialize all the data and deserialize it again.

https://sequel.jeremyevans.net/rdoc/files/doc/sql_rdoc.html states:

> [Sequel](https://sequel.jeremyevans.net/rdoc/classes/Sequel.html) also supports the SQL EXISTS operator using Dataset#exists:
>
> DB[:albums].exists # EXISTS (SELECT * FROM albums)

So I'd expect this to work
```suggestion
        end.exist
```

And then also remove the negation above, but multi line doesn't work.

Reply:
Exist was returning a query rather than a Boolean, so I changed to empty?.

Reply:
Found out  that Sequel is smart with the `empty?` method in the midst of refactoring, so I'm going to keep it as-is: https://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html#method-i-empty-3F